src/time_tracker/invoices/core.clj

```diff
@@ -1,49 +1,149 @@
+            [clj-time.core :as time]
+            [clj-time.format :as time-format]
+            [clojure.spec :as s]
+            [time-tracker.invoices.spec :as invoices-spec]
+            [time-tracker.spec :as core-spec]))
 
 (defn- seconds->hours
   [seconds]
   (double (/ seconds 3600)))
 
-(defn- round-to-places
-  "Rounds a floating point number to `places` decimal places."
-  [number places]
-  (.setScale (bigdec number) places java.math.BigDecimal/ROUND_HALF_UP))
-
 (defn add-hours
   "Adds the hours logged against the timer entry
   to the `user-id->hours` map."
   [user-id->hours {:keys [app-user-id] :as timer}]
   (let [hours-to-add  (-> (timers-core/elapsed-time timer)
                           (seconds->hours)
-                          (round-to-places 4))
+                          (util/round-to-two-places))
         current-hours (get user-id->hours app-user-id)
-        new-hours     (+ (or current-hours 0) hours-to-add)]
+        new-hours     (+ (or current-hours 0.00M) hours-to-add)]
     (assoc user-id->hours app-user-id new-hours)))
 
 (defn build-user-id->hours
   "Returns a map of {user-id hours-logged}"
   [required-user-ids timers]
-  (let [user-id->zero (zipmap required-user-ids (repeat 0))]
+  (let [user-id->zero (zipmap required-user-ids (repeat 0.00M))]
     (reduce add-hours user-id->zero (vals timers))))
 
-(defn id->name [normalized-entities]
-  (fmap :name normalized-entities))
+(defn- user-hours
+  "Generates a seq of {:id :rate :hours}.
+  There will be a row for every user in `users`."
+  [users timers user-id->rate-vector]
+  (let [user-ids       (keys users)
+        user-id->hours (build-user-id->hours user-ids timers)
+        user-id->rate (into {} (map #(hash-map (:user-id %) (:rate %)) user-id->rate-vector))]
+    (for [user-id user-ids]
+      {:id    user-id
+       :rate  (user-id->rate user-id)
+       :hours (user-id->hours user-id)})))
 
-(defn csv-rows
-  [user-id->hours user-id->name]
-  (into [] (util/transform-keys user-id->hours user-id->name)))
+(defn invoice
+  "Constructs an invoice"
+  [{:keys [users timers user-id->rate] :as invoice-params}]
+  {:pre  [(seq users)]
+   :post [(s/valid? ::invoices-spec/invoice %)]}
+  (merge (select-keys invoice-params [:start :end :client :address
+                                      :notes :tax-rates :currency :utc-offset])
+         {:user-hours (user-hours users timers user-id->rate)}))
 
-(defn generate-csv
-  "Generates a CSV given `users` and `timers`.
-  There will be a row for every user in `users`."
-  [users timers]
-  (let [user-id->name  (id->name users)
-        user-id->hours (build-user-id->hours (keys users) timers)]
-    (with-out-str
-      (csv/write-csv *out*
-                     (->> (csv-rows user-id->hours user-id->name)
-                          (cons ["Name" "Hours Logged"]))))))
+;; TODO: Map these to their symbols and render ₹ properly
+(def currency-symbols
+  {:usd "USD "
+   :inr "INR "})
```

Comment:
is the trailing space here intentional? if so, it's prolly better positioned at #62.